File: cricket-api/functions/matches_scorecard.js
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs').promises;

// Add delay function to avoid overwhelming the server
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

async function scrapeMatches() {
    console.log("Scraping matches data...");
    
    try {
        const link = "https://www.cricbuzz.com/cricket-match/live-scores";
        const response = await axios.get(link, { timeout: 10000 });
        const $ = cheerio.load(response.data);

        const container = $(".cb-col.cb-col-100.cb-bg-white");
        if (container.length === 0) {
            console.log("No container found for matches");
            return [];
        }
            
        const matches = container.find(".cb-mtch-lst.cb-col.cb-col-100.cb-tms-itm");
        
        // Initialize arrays
        const match = [];
        const matchNo = [];
        const dateStadium = [];
        const live = [];
        const scorecard = [];
        const fullCommentary = [];
        const teams = [];

        // Extract match titles
        matches.each((index, card) => {
            try {
                const title = $(card).find("a.text-hvr-underline.text-bold");
                match.push(title.length ? title.text().trim() : "");
            } catch (e) {
                console.log(`Error extracting match title: ${e.message}`);
                match.push("");
            }
        });

        // Extract match numbers
        matches.each((index, card) => {
            try {
                const title = $(card).find("span.text-gray");
                if (title.length) {
                    const text = title.text().trim().replace(/&nbsp;\S*/g, '');
                    matchNo.push(text);
                } else {
                    matchNo.push("");
                }
            } catch (e) {
                console.log(`Error extracting match number: ${e.message}`);
                matchNo.push("");
            }
        });

        // Extract date and stadium
        matches.each((index, card) => {
            try {
                const title = $(card).find("div.text-gray");
                if (title.length) {
                    const text = title.text().trim().replace(/\xa0\S*/g, '');
                    dateStadium.push(text);
                } else {
                    dateStadium.push("");
                }
            } catch (e) {
                console.log(`Error extracting date/stadium: ${e.message}`);
                dateStadium.push("");
            }
        });

        // Extract live score links
        matches.each((index, card) => {
            try {
                const title = $(card).find('a[title="Live Score"]');
                if (title.length) {
                    const text = title.text().trim();
                    const href = title.attr("href") || "";
                    live.push({ text: text, url: href });
                } else {
                    live.push({ text: "Live Score", url: "" });
                }
            } catch (e) {
                console.log(`Error extracting live score link: ${e.message}`);
                live.push({ text: "Live Score", url: "" });
            }
        });

        // Extract scorecard links
        matches.each((index, card) => {
            try {
                const title = $(card).find('a[title="Scorecard"]');
                if (title.length) {
                    const text = title.text().trim();
                    const href = title.attr("href") || "";
                    scorecard.push({ text: text, url: href });
                } else {
                    scorecard.push({ text: "Scorecard", url: "" });
                }
            } catch (e) {
                console.log(`Error extracting scorecard link: ${e.message}`);
                scorecard.push({ text: "Scorecard", url: "" });
            }
        });

        // Extract full commentary links
        matches.each((index, card) => {
            try {
                const title = $(card).find('a[title="Full Commentary"]');
                if (title.length) {
                    const text = title.text().trim();
                    const href = title.attr("href") || "";
                    fullCommentary.push({ text: text, url: href });
                } else {
                    fullCommentary.push({ text: "Full Commentary", url: "" });
                }
            } catch (e) {
                console.log(`Error extracting commentary link: ${e.message}`);
                fullCommentary.push({ text: "Full Commentary", url: "" });
            }
        });

        // Extract team information
        matches.each((index, card) => {
            try {
                // Match status logic
                const liveStatus = $(card).find(".cb-text-live");
                const completeStatus = $(card).find(".cb-text-complete");
                const previewStatus = $(card).find("span.cb-text-preview");
                
                let matchStatus = "";
                if (liveStatus.length) {
                    matchStatus = liveStatus.text().trim();
                } else if (completeStatus.length) {
                    matchStatus = completeStatus.text().trim();
                } else if (previewStatus.length) {
                    matchStatus = previewStatus.text().trim();
                }

                const allCbOvrFlo = $(card).find(".cb-ovr-flo");

                const team1 = allCbOvrFlo.length > 1 ? $(allCbOvrFlo[1]).text().trim() : "";
                const team2 = allCbOvrFlo.length > 3 ? $(allCbOvrFlo[3]).text().trim() : "";
                const score1 = allCbOvrFlo.length > 2 ? $(allCbOvrFlo[2]).text().trim() : "";
                const score2 = allCbOvrFlo.length > 4 ? $(allCbOvrFlo[4]).text().trim() : "";

                const team = {
                    team1: team1,
                    team2: team2,
                    score1: score1,
                    score2: score2,
                    match_status: matchStatus
                };
                teams.push(team);
                
            } catch (e) {
                console.log(`Error extracting team info: ${e.message}`);
                teams.push({
                    team1: "",
                    team2: "",
                    score1: "",
                    score2: "",
                    match_status: ""
                });
            }
        });

        // Combine all data
        const result = [];
        const maxLength = Math.max(match.length, matchNo.length, dateStadium.length, live.length, scorecard.length, fullCommentary.length, teams.length);
        
        for (let i = 0; i < maxLength; i++) {
            result.push({
                match: i < match.length ? match[i] : "",
                status: i < matchNo.length ? matchNo[i] : "",
                date_stadium: i < dateStadium.length ? dateStadium[i] : "",
                live_score: i < live.length ? live[i].url : "",
                scorecard_links: i < scorecard.length ? scorecard[i].url : "",
                commentary: i < fullCommentary.length ? fullCommentary[i].url : "",
                teams: i < teams.length ? teams[i] : {
                    team1: "",
                    team2: "",
                    score1: "",
                    score2: "",
                    match_status: ""
                }
            });
        }

        return result;
        
    } catch (error) {
        console.log(`Error in scrapeMatches: ${error.message}`);
        return [];
    }
}

async function scrapeScorecard(matchesData) {
    console.log("Scraping scorecard data...");
    const scorecard = [];
    
    for (const match of matchesData) {
        if (!match.scorecard_links) {
            console.log(`No scorecard link for match: ${match.match || 'Unknown'}`);
            continue;
        }
            
        try {
            const scorecardLink = "https://www.cricbuzz.com" + match.scorecard_links;
            console.log(`Processing scorecard for: ${scorecardLink}`);
            
            const response = await axios.get(scorecardLink, { timeout: 10000 });
            const $ = cheerio.load(response.data);

            // Create a new match object for each match
            const matchData = {
                match: "",
                innings: []
            };

            // Get match name from page title or header
            const matchTitle = $(".cb-nav-hdr.cb-font-18.line-ht24");
            if (matchTitle.length) {
                matchData.match = matchTitle.text().trim().split('-')[0];
                console.log(`Processing match: ${matchData.match}`);
            } else {
                matchData.match = match.match || "Unknown Match";
            }
       
            // Get all innings divs
            const inningseDivs = $('[id^="innings_"]');

            inningseDivs.each((index, inningsDiv) => {
                try {
                    const $inningsDiv = $(inningsDiv);
                    
                    // Get team header
                    const teamHeader = $inningsDiv.find('.cb-scrd-hdr-rw span');
                    const inningsName = teamHeader.length ? teamHeader.text().trim() : "Unknown Team";

                    const inningsData = {
                        innings_name: inningsName,
                        batters: [],
                        extras: "",
                        total: "",
                        yet_to_bat: [],
                        fall_of_wickets: [],
                        bowlers: []
                    };
                    
                    // Get all items in this innings
                    const allItems = $inningsDiv.find('.cb-scrd-itms');
                    
                    // Process batters data
                    allItems.each((itemIndex, item) => {
                        try {
                            const $item = $(item);
                            const nameDiv = $item.find('.cb-col-25');
                            const nameLink = nameDiv.find('a');
                            
                            if (nameLink.length) {
                                const batterName = nameLink.text().trim();
                                
                                // Skip if it's an extras or total row
                                if (batterName === "Extras" || batterName === "Total") {
                                    return;
                                }
                                    
                                const statusDiv = $item.find('.cb-col-33');
                                const runsDiv = $item.find('.cb-col-8.text-right.text-bold');
                                
                                // Get all text-right columns for balls, 4s, 6s, SR
                                const textRightCols = $item.find('.cb-col-8.text-right');
                                const ballsDiv = textRightCols.length > 1 ? $(textRightCols[1]) : null;
                                const foursDiv = textRightCols.length > 2 ? $(textRightCols[2]) : null;
                                const sixesDiv = textRightCols.length > 3 ? $(textRightCols[3]) : null;
                                const strikeRateDiv = textRightCols.length > 4 ? $(textRightCols[4]) : null;
                                
                                // Clean batter name (remove parentheses content)
                                const cleanName = batterName.replace(/\s*\([^)]*\)/g, '').trim();

                                const batterData = {
                                    name: cleanName,
                                    status: statusDiv.length ? statusDiv.text().trim() : "",
                                    runs: runsDiv.length ? runsDiv.text().trim() : "",
                                    balls: ballsDiv ? ballsDiv.text().trim() : "",
                                    "4s": foursDiv ? foursDiv.text().trim() : "",
                                    "6s": sixesDiv ? sixesDiv.text().trim() : "",
                                    strike_rate: strikeRateDiv ? strikeRateDiv.text().trim() : ""
                                };
                                inningsData.batters.push(batterData);
                            }
                        } catch (e) {
                            console.log(`Error processing batter: ${e.message}`);
                        }
                    });
                    
                    // Process extras
                    allItems.each((itemIndex, item) => {
                        try {
                            const $item = $(item);
                            const extraRunsDiv = $item.find('.cb-col.cb-col-8.text-bold.cb-text-black.text-right');
                            if (extraRunsDiv.length) {
                                const extraRuns = extraRunsDiv.text().trim();
                                const extrasDivision = $item.find('.cb-col-32.cb-col');
                                const extrasDivisionText = extrasDivision.length ? extrasDivision.text().trim() : "";
                                inningsData.extras = extraRuns + extrasDivisionText;
                                return false; // Break out of loop
                            }
                        } catch (e) {
                            console.log(`Error processing extras: ${e.message}`);
                        }
                    });
                    
                    // Process total
                    allItems.each((itemIndex, item) => {
                        try {
                            const $item = $(item);
                            const totalRunsDiv = $item.find('.cb-col.cb-col-8.text-bold.text-black.text-right');
                            if (totalRunsDiv.length) {
                                const totalRuns = totalRunsDiv.text().trim();
                                const totalDivision = $item.find('.cb-col-32.cb-col');
                                const totalDivisionText = totalDivision.length ? totalDivision.text().trim() : "";
                                inningsData.total = totalRuns + totalDivisionText;
                                return false; // Break out of loop
                            }
                        } catch (e) {
                            console.log(`Error processing total: ${e.message}`);
                        }
                    });
                    
                    // Process yet to bat players
                    allItems.each((itemIndex, item) => {
                        try {
                            const $item = $(item);
                            const yetToBatSection = $item.find('.cb-col-73.cb-col');
                            yetToBatSection.each((playerIndex, player) => {
                                const playerText = $(player).text().trim();
                                if (playerText && playerText !== "Yet to Bat") {
                                    inningsData.yet_to_bat.push(playerText);
                                }
                            });
                        } catch (e) {
                            console.log(`Error processing yet to bat: ${e.message}`);
                        }
                    });
                    
                    // Process fall of wickets
                    try {
                        const fallOfWickets = $inningsDiv.find(".cb-col.cb-col-100.cb-col-rt.cb-font-13");
                        fallOfWickets.each((wicketIndex, wicket) => {
                            const wicketText = $(wicket).text().trim();
                            if (wicketText) {
                                inningsData.fall_of_wickets.push(wicketText);
                            }
                        });
                    } catch (e) {
                        console.log(`Error processing fall of wickets: ${e.message}`);
                    }
                    
                    // Process bowlers data
                    allItems.each((itemIndex, item) => {
                        try {
                            const $item = $(item);
                            const bowlerDiv = $item.find('.cb-col-38');
                            const bowlerLink = bowlerDiv.find('a');
                            
                            if (bowlerLink.length) {
                                const bowlerName = bowlerLink.text().trim();
                                
                                // Skip if it's an extras or total row
                                if (bowlerName === "Extras" || bowlerName === "Total") {
                                    return;
                                }
                                
                                // Get bowler statistics with more specific selectors
                                const textRightCols = $item.find('.cb-col-8.text-right');
                                const textRight10Cols = $item.find('.cb-col-10.text-right');
                                
                                const bowlerData = {
                                    bowler_name: bowlerName,
                                    overs: textRightCols.length > 0 ? $(textRightCols[0]).text().trim() : "",
                                    maiden: textRightCols.length > 1 ? $(textRightCols[1]).text().trim() : "",
                                    runs: textRight10Cols.length > 0 ? $(textRight10Cols[0]).text().trim() : "",
                                    wickets: textRightCols.length > 2 ? $(textRightCols[2]).text().trim() : "",
                                    no_balls: textRightCols.length > 3 ? $(textRightCols[3]).text().trim() : "",
                                    wide_balls: textRightCols.length > 4 ? $(textRightCols[4]).text().trim() : "",
                                    economy: textRight10Cols.length > 1 ? $(textRight10Cols[1]).text().trim() : ""
                                };
                                inningsData.bowlers.push(bowlerData);
                            }
                        } catch (e) {
                            console.log(`Error processing bowler: ${e.message}`);
                        }
                    });

                    // Add this innings to the current match
                    matchData.innings.push(inningsData);
                    
                } catch (e) {
                    console.log(`Error processing innings: ${e.message}`);
                }
            });

            // Add the completed match to the scorecard list
            scorecard.push(matchData);
            
            // Add small delay between requests
            await delay(1000);
            
        } catch (error) {
            console.log(`Error processing scorecard for match ${match.match || 'Unknown'}: ${error.message}`);
            // Add empty match data to maintain consistency
            scorecard.push({
                match: match.match || "Unknown Match",
                innings: []
            });
        }
    }

    return scorecard;
}

async function main() {
    console.log("Starting cricket data scraping...");
    
    try {
        // Scrape matches data
        console.log("Scraping matches data...");
        const matchesData = await scrapeMatches();
        
        if (!matchesData || matchesData.length === 0) {
            console.log("No matches data found");
            return;
        }
        
        // Save matches data
        try {
            await fs.writeFile("data/matches.json", JSON.stringify(matchesData, null, 2), 'utf8');
            console.log(`Saved ${matchesData.length} matches to matches.json`);
        } catch (e) {
            console.log(`Error saving matches data: ${e.message}`);
            return;
        }
        
        // Scrape scorecard data
        console.log("Scraping scorecard data...");
        const scorecardData = await scrapeScorecard(matchesData);
        
        // Save scorecard data
        try {
            await fs.writeFile("data/scorecard.json", JSON.stringify(scorecardData, null, 2), 'utf8');
            console.log(`Saved ${scorecardData.length} scorecards to scorecard.json`);
        } catch (e) {
            console.log(`Error saving scorecard data: ${e.message}`);
        }
        
        console.log("Cricket data scraping completed!");
        
    } catch (error) {
        console.log(`Error in main function: ${error.message}`);
    }
}

// Run the scraper
if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    scrapeMatches,
    scrapeScorecard,
    main
};

File: cricket-live-data-python/matches_scorecard.py

import re
import requests
from bs4 import BeautifulSoup
import json
import time

def scrape_matches():
    """Scrape live cricket matches data"""
    try:
        link = "https://www.cricbuzz.com/cricket-match/live-scores"
        source = requests.get(link, timeout=10).text
        page = BeautifulSoup(source, "lxml")

        container = page.find("div", class_="cb-col cb-col-100 cb-bg-white")
        if not container:
            print("No container found for matches")
            return []
            
        matches = container.find_all("div", class_="cb-mtch-lst cb-col cb-col-100 cb-tms-itm")
        
        # Initialize lists
        match = []
        match_no = []
        date_stadium = []
        live = []
        scorecard = []
        FullCommentary = []
        teams = []

        # Extract match titles
        for card in matches:
            try:
                title = card.find("a", class_="text-hvr-underline text-bold")
                match.append(title.text.strip() if title else "")
            except Exception as e:
                print(f"Error extracting match title: {e}")
                match.append("")

        # Extract match numbers
        for card in matches:
            try:
                title = card.find("span", class_="text-gray")
                if title:
                    match_no.append(re.sub(r'\&nbsp;\S*', '', title.text.strip()))
                else:
                    match_no.append("")
            except Exception as e:
                print(f"Error extracting match number: {e}")
                match_no.append("")

        # Extract date and stadium
        for card in matches:
            try:
                title = card.find("div", class_="text-gray")
                if title:
                    date_stadium.append(re.sub(r'\xa0\S*', '', title.text.strip()))
                else:
                    date_stadium.append("")
            except Exception as e:
                print(f"Error extracting date/stadium: {e}")
                date_stadium.append("")

        # Extract live score links
        for card in matches:
            try:
                title = card.find("a", attrs={"title": "Live Score"})
                if title:
                    text = title.text.strip()
                    href = title.get("href", "")
                    live.append({"text": text, "url": href})
                else:
                    live.append({"text": "Live Score", "url": ""})
            except Exception as e:
                print(f"Error extracting live score link: {e}")
                live.append({"text": "Live Score", "url": ""})

        # Extract scorecard links
        for card in matches:
            try:
                title = card.find("a", attrs={"title": "Scorecard"})
                if title:
                    text = title.text.strip()
                    href = title.get("href", "")
                    scorecard.append({"text": text, "url": href})
                else:
                    scorecard.append({"text": "Scorecard", "url": ""})
            except Exception as e:
                print(f"Error extracting scorecard link: {e}")
                scorecard.append({"text": "Scorecard", "url": ""})

        # Extract full commentary links
        for card in matches:
            try:
                title = card.find("a", attrs={"title": "Full Commentary"})
                if title:
                    text = title.text.strip()
                    href = title.get("href", "")
                    FullCommentary.append({"text": text, "url": href})
                else:
                    FullCommentary.append({"text": "Full Commentary", "url": ""})
            except Exception as e:
                print(f"Error extracting commentary link: {e}")
                FullCommentary.append({"text": "Full Commentary", "url": ""})

        # Extract team information
        for card in matches:
            try:
                # Match status logic
                match_status = (card.find("div", class_="cb-text-live") or 
                              card.find("div", class_="cb-text-complete") or 
                              card.find("span", class_="cb-text-preview"))
                match_status = match_status.text.strip() if match_status else ""

                all_cb_ovr_flo = card.find_all("div", class_="cb-ovr-flo")

                team1 = all_cb_ovr_flo[1].text.strip() if len(all_cb_ovr_flo) > 1 else ""
                team2 = all_cb_ovr_flo[3].text.strip() if len(all_cb_ovr_flo) > 3 else ""
                score1 = all_cb_ovr_flo[2].text.strip() if len(all_cb_ovr_flo) > 2 else ""
                score2 = all_cb_ovr_flo[4].text.strip() if len(all_cb_ovr_flo) > 4 else ""

                team = {
                    "team1": team1,
                    "team2": team2,
                    "score1": score1,
                    "score2": score2,
                    "match_status": match_status
                }
                teams.append(team)
                
            except Exception as e:
                print(f"Error extracting team info: {e}")
                teams.append({
                    "team1": "",
                    "team2": "",
                    "score1": "",
                    "score2": "",
                    "match_status": ""
                })

        # Combine all data
        result = []
        max_length = max(len(match), len(match_no), len(date_stadium), len(live), len(scorecard), len(FullCommentary), len(teams))
        
        for i in range(max_length):
            result.append({
                "match": match[i] if i < len(match) else "",
                "status": match_no[i] if i < len(match_no) else "",
                "date_stadium": date_stadium[i] if i < len(date_stadium) else "",
                "live_score": live[i]['url'] if i < len(live) else "",
                "scorecard_links": scorecard[i]['url'] if i < len(scorecard) else "",
                "commentary": FullCommentary[i]['url'] if i < len(FullCommentary) else "",
                "teams": teams[i] if i < len(teams) else {
                    "team1": "",
                    "team2": "",
                    "score1": "",
                    "score2": "",
                    "match_status": ""
                }
            })

        return result
        
    except Exception as e:
        print(f"Error in scrape_matches: {e}")
        return []

def scrape_scorecard(matches_data):
    """Scrape scorecard data for all matches"""
    scorecard = []
    
    for match in matches_data:
        if not match.get("scorecard_links"):
            print(f"No scorecard link for match: {match.get('match', 'Unknown')}")
            continue
            
        try:
            scorecard_link = "https://www.cricbuzz.com" + match.get("scorecard_links")
            print(f"Processing scorecard for: {scorecard_link}")
            
            source = requests.get(scorecard_link, timeout=10).text
            page = BeautifulSoup(source, "lxml")

            # Create a new match object for each match
            match_data = {
                "match": "",
                "innings": []
            }

            # Get match name from page title or header
            match_title = page.select_one('.cb-nav-hdr.cb-font-18.line-ht24')
            if match_title:
                match_data["match"] = match_title.text.strip().split('-')[0]
                print(f"Processing match: {match_data['match']}")
            else:
                match_data["match"] = match.get("match", "Unknown Match")
       
            # Get all innings divs
            innings_divs = page.select('[id^="innings_"]')

            for innings_div in innings_divs:
                try:
                    # Get team header
                    team_header = innings_div.select_one('.cb-scrd-hdr-rw span')
                    innings_name = team_header.text.strip() if team_header else "Unknown Team"

                    innings_data = {
                        "innings_name": innings_name,
                        "batters": [],
                        "extras": "",
                        "total": "",
                        "yet_to_bat": [],
                        "fall_of_wickets": [],
                        "bowlers": []
                    }
                    
                    # Get all items in this innings
                    all_items = innings_div.select('.cb-scrd-itms')
                    
                    # Process batters data
                    for item in all_items:
                        try:
                            name_div = item.select_one('.cb-col-25')
                            if name_div and name_div.a:
                                batter_name = name_div.a.text.strip()
                                
                                # Skip if it's an extras or total row
                                if batter_name in ["Extras", "Total"]:
                                    continue
                                    
                                status_div = item.select_one('.cb-col-33')
                                runs_div = item.select_one('.cb-col-8.text-right.text-bold')
                                
                                # Get all text-right columns for balls, 4s, 6s, SR
                                text_right_cols = item.select('.cb-col-8.text-right')
                                balls_div = text_right_cols[1] if len(text_right_cols) > 1 else None
                                fours_div = text_right_cols[2] if len(text_right_cols) > 2 else None
                                sixes_div = text_right_cols[3] if len(text_right_cols) > 3 else None
                                strike_rate_div = text_right_cols[4] if len(text_right_cols) > 4 else None
                                
                                # Clean batter name (remove parentheses content)
                                clean_name = re.sub(r'\s*\([^)]*\)', '', batter_name).strip()

                                batter_data = {
                                    "name": clean_name,
                                    "status": status_div.text.strip() if status_div else "",
                                    "runs": runs_div.text.strip() if runs_div else "",
                                    "balls": balls_div.text.strip() if balls_div else "",
                                    "4s": fours_div.text.strip() if fours_div else "",
                                    "6s": sixes_div.text.strip() if sixes_div else "",
                                    "strike_rate": strike_rate_div.text.strip() if strike_rate_div else ""
                                }
                                innings_data["batters"].append(batter_data)
                        except Exception as e:
                            print(f"Error processing batter: {e}")
                            continue
                    
                    # Process extras
                    for item in all_items:
                        try:
                            if item.select_one('.cb-col.cb-col-8.text-bold.cb-text-black.text-right'):
                                extra_runs = item.select_one('.cb-col.cb-col-8.text-bold.cb-text-black.text-right').text.strip()
                                extras_division = item.select_one('.cb-col-32.cb-col')
                                extras_division_text = extras_division.text.strip() if extras_division else ""
                                innings_data["extras"] = extra_runs + extras_division_text
                                break
                        except Exception as e:
                            print(f"Error processing extras: {e}")
                            continue
                    
                    # Process total
                    for item in all_items:
                        try:
                            if item.select_one('.cb-col.cb-col-8.text-bold.text-black.text-right'):
                                total_runs = item.select_one('.cb-col.cb-col-8.text-bold.text-black.text-right').text.strip()
                                total_division = item.select_one('.cb-col-32.cb-col')
                                total_division_text = total_division.text.strip() if total_division else ""
                                innings_data["total"] = total_runs + total_division_text
                                break
                        except Exception as e:
                            print(f"Error processing total: {e}")
                            continue
                    
                    # Process yet to bat players
                    for item in all_items:
                        try:
                            yet_to_bat_section = item.select('.cb-col-73.cb-col')
                            for player in yet_to_bat_section:
                                player_text = player.text.strip()
                                if player_text and player_text != "Yet to Bat":
                                    innings_data["yet_to_bat"].append(player_text)
                        except Exception as e:
                            print(f"Error processing yet to bat: {e}")
                            continue
                    
                    # Process fall of wickets
                    try:
                        fall_of_wickets = innings_div.select(".cb-col.cb-col-100.cb-col-rt.cb-font-13")
                        for wicket in fall_of_wickets:
                            wicket_text = wicket.text.strip()
                            if wicket_text:
                                innings_data["fall_of_wickets"].append(wicket_text)
                    except Exception as e:
                        print(f"Error processing fall of wickets: {e}")
                    
                    # Process bowlers data
                    for item in all_items:
                        try:
                            bowler_div = item.select_one('.cb-col-38')
                            if bowler_div and bowler_div.a:
                                bowler_name = bowler_div.a.text.strip()
                                
                                # Skip if it's an extras or total row
                                if bowler_name in ["Extras", "Total"]:
                                    continue
                                
                                # Get bowler statistics with more specific selectors
                                text_right_cols = item.select('.cb-col-8.text-right')
                                text_right_10_cols = item.select('.cb-col-10.text-right')
                                
                                bowler_data = {
                                    "bowler_name": bowler_name,
                                    "overs": text_right_cols[0].text.strip() if len(text_right_cols) > 0 else "",
                                    "maiden": text_right_cols[1].text.strip() if len(text_right_cols) > 1 else "",
                                    "runs": text_right_10_cols[0].text.strip() if len(text_right_10_cols) > 0 else "",
                                    "wickets": text_right_cols[2].text.strip() if len(text_right_cols) > 2 else "",
                                    "no_balls": text_right_cols[3].text.strip() if len(text_right_cols) > 3 else "",
                                    "wide_balls": text_right_cols[4].text.strip() if len(text_right_cols) > 4 else "",
                                    "economy": text_right_10_cols[1].text.strip() if len(text_right_10_cols) > 1 else ""
                                }
                                innings_data["bowlers"].append(bowler_data)
                        except Exception as e:
                            print(f"Error processing bowler: {e}")
                            continue

                    # Add this innings to the current match
                    match_data["innings"].append(innings_data)
                    
                except Exception as e:
                    print(f"Error processing innings: {e}")
                    continue

            # Add the completed match to the scorecard list
            scorecard.append(match_data)
            
            # Add small delay between requests
            time.sleep(1)
            
        except Exception as e:
            print(f"Error processing scorecard for match {match.get('match', 'Unknown')}: {e}")
            # Add empty match data to maintain consistency
            scorecard.append({
                "match": match.get("match", "Unknown Match"),
                "innings": []
            })
            continue

    return scorecard

def main():
    """Main function to scrape matches and scorecard data"""
    print("Starting cricket data scraping...")
    
    # Scrape matches data
    print("Scraping matches data...")
    matches_data = scrape_matches()
    
    if not matches_data:
        print("No matches data found")
        return
    
    # Save matches data
    try:
        with open("matches.json", "w", encoding='utf-8') as f:
            json.dump(matches_data, f, indent=2, ensure_ascii=False)
        print(f"Saved {len(matches_data)} matches to matches.json")
    except Exception as e:
        print(f"Error saving matches data: {e}")
        return
    
    # Scrape scorecard data
    print("Scraping scorecard data...")
    scorecard_data = scrape_scorecard(matches_data)
    
    # Save scorecard data
    try:
        with open("scorecard.json", "w", encoding='utf-8') as f:
            json.dump(scorecard_data, f, indent=2, ensure_ascii=False)
        print(f"Saved {len(scorecard_data)} scorecards to scorecard.json")
    except Exception as e:
        print(f"Error saving scorecard data: {e}")
    
    print("Cricket data scraping completed!")

if __name__ == "__main__":
    main()